### DISCUSSION TITLE
Phalcon 5.0 Future

### DISCUSSION THREAD

**Discussion by ghost** [phalcon/phalcon#96] [General]
Everyone please stop pulling Phalcon into other frameworks and theme engines. Please keep the Phalcon we all love as much compatible with last Phalcon v4.1.0 as possible. Do not change or implement too many features! This should be transitional and backward compatible release only from version 4.1 to 5.0 Phalcon programmers can decide in the future if they want to drasticly change Phalcon in PHP space somehwere in versions 6 or 7. But for those of us who still use Phalcon 4.1 (extension) we need something reliable and easy to switch to when we upgrade. Not having to completely rewrite our projects that we worked hard on all this years. I really hope Phalcon developers will be smart enougt to not be pulled into other frameworks. Otherwise, we just all switch to something else at that point. Thank you all for listening.

**Comment by lajosbencz** [comment]
I disagree.
In the long run, it doesn't matter if v5 or v6 is the one that's backwards incompatible, and since v5 will be rehauled, it makes perfect sense to drop compatibility.
The framework will _probably_ lose performance by dropping the extension distribution, so if that's the main reason you were using it, you should probably stick with v4.
Also, AFAIK, having full (future) PSR compatibility means they MUST break the v4 API.

**Reply by niden** [reply]
> Also, AFAIK, having full (future) PSR compatibility means they MUST break the v4 API.

In most cases this will not be true, since v4 is using PSR interfaces for the Logger, Cache etc. This will be true for middleware for sure and event handlers.

**Comment by mesaglam** [comment]
The only reason we use Phalcon in our SaaS is because it performs perfectly by running as an extension and using fewer system resources. 

Please don't consider dropping extension distribution.

**Comment by wurst-hans** [comment]
I'd love to see a port of extension to pure PHP solution. We have customers using shared hosting were our projects cannot be installed else.

**Comment by betrixed** [comment]
I plead guilty to inter-framework code mixing and borrowing. Mutational breeding. I don't particularly care what people think, since it is for my own education. Of course I do support the concept of a canonical PHP version of phalcon, but it isn't here yet.
 I burrowed bits of phalcon framework for PHP, that I already depended on, to keep up with the times,  and incorporated them into my own mongrel framework, while trying to keep the borrowing to a minimum, to where phalcon already appears to have the best ideas.   In doing so I have learned more of how it is designed to work, because I have spend more time reading and stepping through the code. By modification and playing around and testing.  I best see how things work, by having a live web framework, which I already had.

 I like how codeception works on acceptance and functional testing, I will get to use it more.  I have yet no idea how to write tests for pages that use AJAX get and post and update.

Performance difference between all PHP vs total framework C-extension  is greatly reduced  by PHP 8.0 opcache and jit.  This moves the balance between C-code and interpreted code, in favour of more interpreted code. Framework C-extensions get to be needed for only particular bottle-necks with special performace requirements. All non-framework PHP C-extensions still needed work with the server OS level external world. 

 I have looked at the complexity, of framework components from other sources, and they kind of provide a history of similar art. It is nice to see how ideas and techniques are reused and burrowed and expressed by various other projects.
  I now have a much better idea how the phalcon classes work, even though I replaced some with a simpler expression of some of them, for my purposes. Being modular, should be always kept as a possibility. All those traits and  interfaces seem a little over pedantic, scattered as they are everywhere in the namespace tree.  I much prever to have traits in one folder, and interfaces in another.   I dislike having the long name Interface at the end. I much prefer an 'I' in front of a short name. I even decided that the namespace for interfaces should be the shortend 'Face' .   'Inter' is not necessary.  I am never going to use the entire phalcon framework as is.  I liked phalcon some more after understanding why it seems to be designed the way it is, and why PHP works the way it is.  Even when I have differing preferences to standard, I might change later.
  
   I have revived some of my C coding, to play around with a C-extension for ideas I wanted to explore, after abandoning PHP-CPP, as not keeping up,  to learn the PHP internals better, and following the recent zend extension. I have no idea if I will get to a stage of working with something really useful, and it all takes time.

**Comment by ghost** [comment]
Yeah moving away from Zephir/C to dymaic PHP was a wrong move on Phalcon's team part. It more shows like they were lazy to learn how to continue developing in it in Zephir/C after main programmer left. I'll never use Phalcon v5+ (dynamic) I'll stick to using v4 as long as possible, when that no longer supported will switch to something else. I've used v2 for 5+ years long after it was no longer maintained with a lot of success. v4 is much more mature, so I think it should work for me indefinantly :-) And if the team don't abandon it complete and at least fix bugs in v4 this will be the gratest dessition they have made in the history of phalcon. I mean why would I use PHP in dymaic language with v5 when there is other much better frameworks like Laravel? LOL

**Comment by ghost** [comment]
Here is article @ https://kinsta.com/blog/php-frameworks/
"Phalcon is a PHP framework built for speed. It is delivered as a web server extension written in Zephir and C. No knowledge of C is necessary. Developers work with the PHP classes and namespaces the framework generates."

This was about v4 but this is about to change v5. I think less people will use it as dymaic php. They will just switch over to something better like Laravel instead. Phalcon team should re-think there a lot more and continue learning Zephir/C way or loose Phalcon for good.

Instead of switching to native PHP, phalcon team should ether learn Zephir/C them selfs or at least hire some one who can. I'm sure phalcon community who loved Phlacon up until v4 are willing to donate to make this happen!

**Comment by ghost** [comment]
This might be a good alternative to phalcon now.

https://github.com/laruence/yaf
PHP framework written in c and built as a PHP extension.

If anyone know any other please share. Thank you!

**Comment by scarpelius** [comment]
Sadly, I have to give up on phalcon all together. I've used it for company frontend and it (still) works wonders on high load. But with p5 ditching the extension feature, I am not going to use it anymore. 
Phalcon is becoming yet another framework, loosing its main selling point. 
As a programmer it worth learning and use Symfony because the demand for that is bigger in job interviews.

Thanks for the ride, it was awesome.

**Reply by Jeckerson** [reply]
> But with p5 ditching the extension feature

Version 5 will be as full extension. Version 6 will be PHP + extension as hybrid solution...
We will make post explaining the future of Phalcon. For further information will be in blog post, stay tuned.

**Reply by scarpelius** [reply]
@Jeckerson Please enlighten me: as per this blog post https://blog.phalcon.io/post/the-future-of-phalcon phalcon v5 is no longer a php C extension.
Also, what does it mean "Version 6 will be PHP + extension as hybrid solution".

**Reply by Jeckerson** [reply]
> as per this blog post https://blog.phalcon.io/post/the-future-of-phalcon phalcon v5 is no longer a php C extension.

This is almost 1 year blog post, which we didn't updated due hard working with Zephir ([to make it PHP8 compatible](https://github.com/zephir-lang/zephir/pull/2171)) and now with [Phalcon v5](https://github.com/phalcon/cphalcon/releases/tag/v5.0.0-alpha.1). As I told above, there will be updated post about that.

> "Version 6 will be PHP + extension as hybrid solution".

Why hybrid? Because some things are equal in terms of performance as in PHP...
The solution would be: want to use only in PHP - run `composer require phalcon/phalcon`
Want extra performance with same functionally? - run `pecl install phalcon-ext` or `apt install ...`

Working example of hybrid solution you can find here (read until end) - https://github.com/RubixML/Tensor#installation
Also read the note there:
> Note: If both the library and extension are installed, the extension will take precedence.

**Reply by scarpelius** [reply]
@Jeckerson You guys should make an official statement of the future of Phalcon ASAP.
The blog post I've mentioned earlier is the only info available.

**Comment by Jeckerson** [comment]
Updated blog post about Phalcon future just released - https://blog.phalcon.io/post/phalcon-roadmap

Locking.
